source/slang/slang-lower-to-ir.cpp

```diff
@@ -6107,11 +6120,83 @@ struct DeclLoweringVisitor : DeclVisitor<DeclLoweringVisitor, LoweredValInfo>
             // being accessed, rather than a simple value.
             irResultType = subBuilder->getPtrType(irResultType);
         }
+    }
+
+    IRFuncType* _lowerFuncTypeImpl(
+        ParameterLists& parameterLists,
+        List<IRType*>& paramTypes,
+        IRType*& irResultType,
+        IRBuilder* builder,
+        IRGenContext* irGenContext,
+        FunctionDeclBase* decl)
+    {
+        _lowerFuncResultAndParameterTypes(
+            parameterLists,
+            paramTypes,
+            irResultType,
+            builder,
+            irGenContext,
+            decl);
 
-        auto irFuncType = subBuilder->getFuncType(
+        auto irFuncType = builder->getFuncType(
             paramTypes.getCount(),
             paramTypes.getBuffer(),
             irResultType);
+
+        if (parameterLists.params.getCount() && parameterLists.params[0].isThisParam)
+            builder->addThisPointerDecoration(irFuncType, 0);
+        return irFuncType;
```

Comment:
There is a potential problem with this logic. What if I have:

```
struct S {
    int doA(float x) { ... }
}
int doB(S s, float y) { ... }
```

At the IR level, both `doA` and `doB` would currently lower to use the exact same `IRFuncType`: `(S, y) -> int`.

So what happens if you decorate that single `IRFuncType` with your `IRThisPointerDecoration`? It seems like it will incorrectly flag the `s` parameter of `doB` as a `this` pointer, which isn't correct?

Maybe that isn't actually a problem (I don't fully understand where/how the decoration impacts codegen), but it doesn't seem semantically correct.

Reply:
To follow up on this a bit, we *do* have the starting point of a solution for cases where you want to attach "extra" data to an instruction (like you would do with decorations), but the instruction in question is one that can get hoisted/deduplicated and you don't want the versions with and without your decoration to get mixed up.

There is an `IRAttr` type for "attributes," which are similar to `IRDeocration`s in that they store extra/optional data, but are very different than decorations in that decorations are *children* of the thing they decoration, while attributes are *operands* of the thing they decorate.

You can think of `IRAttr`s as extra operands that can follow the "ordinary" operands of certain instructions (not all instructions are designed to work with attributes today). It is possible to query for the attributes on an instruction, and to search for attributes of a certain type (like you can with deocrations). When creating an instruction of type kind that gets deduplicated, you need to put together both the ordinary operands and all attributes before you construct the instance (you can't go back and add attribute slater), so that the deduplication logic can take the attributes into account (two instructions with the same "ordinary" operands, but different attributes are distinct from a deduplication standpoint).

I don't actually think attributes are a perfect fit for what you are doing in this case (I'm not sure what an ideal design would be) but I wanted to note that they are there in case you ever need to apply them to some challenge.

(E.g., if/when we add a notion of different calling conventions to Slang, we'd have to make a choice of whether to make calling convention a required property of an `IRFuncType` (so that it is always one of the operands), or to make an `IRAttr` that can associate an `IRFuncType` with a non-default calling convention. No matter what, we'd need to ensure that function types with different calling conventions never get confused for one another and/or deduplicated into a single type.)